### PR TITLE
Assign / Remove Trainee by Supervisor

### PR DESCRIPTION
**Pull request by abdullahallmehedi**

### CODE REVIEW

**Comment by zenith19** [comment]
Ok

File: app/models/course.rb [hunk 1]
```diff
@@ -27,6 +27,15 @@ def supervisors
       where("users.supervisor = ? AND courses.id = ?", true, id)
   end
 
+  def assignable_trainees
+    join_sql = "LEFT JOIN users_courses uc ON uc.user_id = users.id"
```
Comment (chuanhtuan) [inline app/models/course.rb]: write subquery in a scope, plz

Comment (srahmanbd) [inline app/models/course.rb]: "assignable_trainees" for this function you need to use scope. :)

File: app/controllers/supervisor/users_courses_controller.rb [hunk 2]
```diff
@@ -22,12 +23,22 @@ def load_users
...
+      @users = @course.assignable_trainees      
     end
   end
 
   def course_params
     params.require(:course).permit user_ids: []
   end
+
+  def check_course
+    if params[:assign_type] == UsersCourse::ASSIGN_TYPE[:supervisor] && @course.finished?
```
Comment (abdullahallmehedi) [inline app/controllers/supervisor/users_courses_controller.rb]: 80

File: app/controllers/supervisor/users_courses_controller.rb [hunk 3]
```diff
@@ -22,12 +23,22 @@ def load_users
...
+      flash[:danger] = t "flash_user_course_supervisor_edit"
+      redirect_to supervisor_course_path(@course)
+    elsif params[:assign_type] == UsersCourse::ASSIGN_TYPE[:trainee] && @course.started?
```
Comment (abdullahallmehedi) [inline app/controllers/supervisor/users_courses_controller.rb]: 80

File: app/models/user.rb [hunk 4]
```diff
@@ -17,10 +17,11 @@ class User < ActiveRecord::Base
   has_many :users_tasks, dependent: :destroy
   has_many :tasks, through: :users_tasks
 
-  scope :supervisors, -> { where(supervisor: true) }
-  scope :trainees, -> { where(supervisor: false) }
+  scope :supervisors, -> { where supervisor: true }
+  scope :trainees, -> { where supervisor: false }
+  scope :course_unfinished,  -> { joins(:courses).where.not courses: {status: Course::STATUS[:finished]} }
```
Comment (abdullahallmehedi) [inline app/models/user.rb]: 80

File: app/models/course.rb [hunk 5]
```diff
@@ -16,15 +16,22 @@ class Course < ActiveRecord::Base
...
   
   def supervisors
-    User.joins(:courses).
-      where("users.supervisor = ? AND courses.id = ?", true, id)
+    users.where(supervisor: true)
+  end
+
+  def assignable_trainees    
+    excluded_trainees_ids = User.course_unfinished.trainees.ids - trainees.ids
+    User.where.not(id: excluded_trainees_ids).trainees
```
Comment (dieunb) [inline app/models/course.rb]: move `where` to `scope` pls

File: app/models/user.rb [hunk 6]
```diff
@@ -17,10 +17,13 @@ class User < ActiveRecord::Base
   has_many :users_tasks, dependent: :destroy
   has_many :tasks, through: :users_tasks
 
-  scope :supervisors, -> { where(supervisor: true) }
-  scope :trainees, -> { where(supervisor: false) }
+  scope :supervisors, -> { where supervisor: true }
```
Comment (dieunb) [inline app/models/user.rb]: convention, should
```
{where supervisor: true}
```

File: app/models/user.rb [hunk 7]
```diff
@@ -17,10 +17,13 @@ class User < ActiveRecord::Base
...
+  scope :trainees, -> { where supervisor: false }
```
Comment (dieunb) [inline app/models/user.rb]: same above

File: app/controllers/supervisor/users_courses_controller.rb [hunk 8]
```diff
@@ -22,12 +23,24 @@ def load_users
...
 
   def course_params
     params.require(:course).permit user_ids: []
   end
+
+  def check_course
+    if params[:assign_type] == UsersCourse::ASSIGN_TYPE[:supervisor] &&
+        @course.finished?
+      flash[:danger] = t "flash_user_course_supervisor_edit"
+      redirect_to supervisor_course_path(@course)
```
Comment (MuhammadTamzid) [inline app/controllers/supervisor/users_courses_controller.rb]: remove ()

File: app/controllers/supervisor/users_courses_controller.rb [hunk 9]
```diff
@@ -22,12 +23,24 @@ def load_users
...
+    elsif params[:assign_type] == UsersCourse::ASSIGN_TYPE[:trainee] && 
+        @course.started?
+      flash[:danger] = t "flash_user_course_trainee_edit"
+      redirect_to supervisor_course_path(@course)
```
Comment (MuhammadTamzid) [inline app/controllers/supervisor/users_courses_controller.rb]: remove ()

File: app/models/user.rb (hunk 6 above)
Comment (abdullahallmehedi) [inline app/models/user.rb]: my code is exactly following the convention you mentioned

File: app/models/user.rb (hunk 7 above)
Comment (abdullahallmehedi) [inline app/models/user.rb]: my code is exactly following the convention you mentioned

File: app/models/course.rb [hunk 10]
```diff
@@ -19,13 +19,11 @@ class Course < ActiveRecord::Base
   after_initialize :default_values
 
   def trainees
-    User.joins(:courses).
-      where("users.supervisor = ? AND courses.id = ?", false, id)
+    users.where(supervisor: false)
```
Comment (abdullahallmehedi) [inline app/models/course.rb]: remove ()

File: app/models/course.rb [hunk 11]
```diff
@@ -19,13 +19,11 @@ class Course < ActiveRecord::Base
...
   end
   
   def supervisors
-    User.joins(:courses).
-      where("users.supervisor = ? AND courses.id = ?", true, id)
+    users.where(supervisor: true)
```
Comment (abdullahallmehedi) [inline app/models/course.rb]: remove ()

File: app/models/user.rb [hunk 12]
```diff
@@ -17,10 +17,20 @@ class User < ActiveRecord::Base
   has_many :users_tasks, dependent: :destroy
   has_many :tasks, through: :users_tasks
 
-  scope :supervisors, -> { where(supervisor: true) }
-  scope :trainees, -> { where(supervisor: false) }
+  scope :supervisors, -> {where supervisor: true}
+  scope :trainees, -> {where supervisor: false}
+  scope :course_unfinished,  -> {
```
Comment (dieunb) [inline app/models/user.rb]: remove, we can implement on the next time.

File: app/models/user.rb [hunk 13]
```diff
@@ -17,10 +17,20 @@ class User < ActiveRecord::Base
...
   has_many :tasks, through: :users_tasks
 
-  scope :supervisors, -> { where(supervisor: true) }
-  scope :trainees, -> { where(supervisor: false) }
+  scope :supervisors, -> {where supervisor: true}
+  scope :trainees, -> {where supervisor: false}
+  scope :course_unfinished,  -> { 
+    joins(:courses).where.not courses: {status: Course::STATUS[:finished]} 
+  }
+  scope :assignable_trainees, ->(course) {
```
Comment (dieunb) [inline app/models/user.rb]: remove, we can implement on the next time.

File: app/models/user.rb [hunk 14]
```diff
@@ -17,10 +17,20 @@ class User < ActiveRecord::Base
...
+    excluded_trainees_ids = course_unfinished.trainees.ids - assigned_to(course).trainees.ids
+    where.not(id: excluded_trainees_ids).trainees
+  }
+  scope :assigned_to, ->(course) {
```
Comment (dieunb) [inline app/models/user.rb]: remove, we can implement on the next time.

**Pull request merged**